# Data Cleaning

By Raj Saluja (11/5/19)

### Introduction

In this lab I explored three datasets to see what problems they had and then I cleaned two of them based on what I found through the exploration. Though there was a minimum requirement, I cleaned these datasets as much as I possibly could to get the most practice. During this lab I analyzed datasets to find problems, corrected bad headers, deleted rows, and deleted columns in order to make my two datasets as clean as possible.

### Dataset Exploration

In [153]:
import math as m
import numpy as np
import scipy as sp
import pandas as pd
import matplotlib as mpl
import matplotlib.pyplot as plt
import seaborn as sns

In [154]:
attendance = pd.read_excel('attendance.xls', encoding = "iso-8859-1")

candy = pd.read_excel('BOING-BOING-CANDY-HIERARCHY-2016-SURVEY-Responses.xlsx', encoding = "iso-8859-1")

sales = pd.read_csv('sales.csv', encoding = "iso-8859-1")

Here I opened the three datasets I am going to be cleaning for this lab. If  hadn't done this I wouldn't have been able to clean the data.

In [155]:
attendance.head()

Table 43. Average daily attendance (ADA) as a percentage of total enrollment, school day length, and school year length in public schools, by school level and state: 2007-08  \
0                                                NaN                                                                                                                              
1                                                NaN                                                                                                                              
2                                                  1                                                                                                                              
3                             United States ........                                                                                                                              
4                          Alabama .................                                                                                                                              

                                          Unnamed: 1  Unnamed: 2  \
0  Total elementary, secondary, and combined elem...         NaN   
1                       ADA as percent of enrollment         NaN   
2                                                  2         NaN   
3                                             93.079    0.218555   
4                                            93.8124    1.237261   

                    Unnamed: 3  Unnamed: 4                   Unnamed: 5  \
0                          NaN         NaN                          NaN   
1  Average hours in school day         NaN  Average days in school year   
2                            3         NaN                            4   
3                       6.6447    0.017650                          180   
4                      7.02852    0.065621                          180   

   Unnamed: 6                    Unnamed: 7  Unnamed: 8  \
0         NaN                           NaN         NaN   
1         NaN  Average hours in school year         NaN   
2         NaN                             5         NaN   
3    0.142993                       1192.65      3.0859   
4    0.755110                       1266.62     12.3343   

                     Unnamed: 9  Unnamed: 10                  Unnamed: 11  \
0            Elementary schools          NaN                          NaN   
1  ADA as percent of enrollment          NaN  Average hours in school day   
2                             6          NaN                            7   
3                        94.005     0.269369                      6.65606   
4                       93.7764     1.839154                      7.03847   

   Unnamed: 12                   Unnamed: 13 Unnamed: 14  \
0          NaN             Secondary schools         NaN   
1          NaN  ADA as percent of enrollment         NaN   
2          NaN                             8         NaN   
3     0.015998                       91.1181    0.432223   
4     0.075944                        94.561     0.37928   

                   Unnamed: 15 Unnamed: 16  
0                          NaN         NaN  
1  Average hours in school day         NaN  
2                            9         NaN  
3                      6.59439    0.040308  
4                      7.13724     0.17312

I printed the top of the attendence data to get a feel of what the dataset it like. I see that there are some serious problems with the headers, as there are three lines of headers which is confusing the code. I also see that there are a lot of NaNs, but those are a result of the messed up headers. I also see that there are 17 columns. Finally, I see what the dataset is about based on the columns. It goes over a ton of school statistics including enrollment, average day length, average number of days, how much school there is total, etc.

In [156]:
attendance.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 68 entries, 0 to 67
Data columns (total 17 columns):
Table 43. Average daily attendance (ADA) as a percentage of total enrollment, school day length, and school year length in public schools, by school level and state: 2007-08    57 non-null object
Unnamed: 1                                                                                                                                                                       55 non-null object
Unnamed: 2                                                                                                                                                                       52 non-null float64
Unnamed: 3                                                                                                                                                                       54 non-null object
Unnamed: 4                                                                                                      

In [157]:
attendance.tail()

Table 43. Average daily attendance (ADA) as a percentage of total enrollment, school day length, and school year length in public schools, by school level and state: 2007-08  \
63                          Wyoming .................                                                                                                                              
64                                   †Not applicable.                                                                                                                              
65    ‡Reporting standards not met (too few cases).                                                                                                                                
66  NOTE: Averages reflect data reported by school...                                                                                                                              
67  SOURCE: U.S. Department of Education, National...                                                                                                                              

   Unnamed: 1  Unnamed: 2 Unnamed: 3  Unnamed: 4 Unnamed: 5  Unnamed: 6  \
63     92.353    1.154487    6.85867    0.045848        175    1.278829   
64        NaN         NaN        NaN         NaN        NaN         NaN   
65        NaN         NaN        NaN         NaN        NaN         NaN   
66        NaN         NaN        NaN         NaN        NaN         NaN   
67        NaN         NaN        NaN         NaN        NaN         NaN   

   Unnamed: 7  Unnamed: 8 Unnamed: 9  Unnamed: 10 Unnamed: 11  Unnamed: 12  \
63    1200.93      8.3266    92.1909     1.647681     6.85061     0.054651   
64        NaN         NaN        NaN          NaN         NaN          NaN   
65        NaN         NaN        NaN          NaN         NaN          NaN   
66        NaN         NaN        NaN          NaN         NaN          NaN   
67        NaN         NaN        NaN          NaN         NaN          NaN   

   Unnamed: 13 Unnamed: 14 Unnamed: 15 Unnamed: 16  
63     92.3894    0.749597     6.97859      0.0679  
64         NaN         NaN         NaN         NaN  
65         NaN         NaN         NaN         NaN  
66         NaN         NaN         NaN         NaN  
67         NaN         NaN         NaN         NaN

Here I checked the tail of my dataset to make sure there weren't anymore problems, but there were. I appears as if there is a final section in which the creators of the dataset made some notes. It also appears as if the symbols "†" and "‡" appear throughout the dataset, but that isn't a problem because the are essentially fill-ins for NaNs and don't make the dataset unclean. Both of these things are problems because they will prevent certain pieces of data from being a number which prevents us from using the dataset.

In [158]:
attendance

Table 43. Average daily attendance (ADA) as a percentage of total enrollment, school day length, and school year length in public schools, by school level and state: 2007-08  \
0                                                 NaN                                                                                                                              
1                                                 NaN                                                                                                                              
2                                                   1                                                                                                                              
3                              United States ........                                                                                                                              
4                           Alabama .................                                                                                                                              
5                           Alaska ..................                                                                                                                              
6                           Arizona .................                                                                                                                              
7                           Arkansas ................                                                                                                                              
8                           California ..............                                                                                                                              
9                                                 NaN                                                                                                                              
10                          Colorado ................                                                                                                                              
11                          Connecticut .............                                                                                                                              
12                          Delaware ................                                                                                                                              
13                          District of Columbia ....                                                                                                                              
14                          Florida .................                                                                                                                              
15                                                NaN                                                                                                                              
16                          Georgia .................                                                                                                                              
17                          Hawaii ..................                                                                                                                              
18                          Idaho ...................                                                                                                                              
19                          Illinois ................                                                                                                                              
20                          Indiana .................                                                                                                                              
21                                         

Here I printed the entire dataset so I could scroll through it and make sure there were no issues in the middle, but there are. It appears that every once in a while there is a row that has no data in it.

In [159]:
candy.head()

Timestamp  \
0 2016-10-24 05:09:23.033   
1 2016-10-24 05:09:54.798   
2 2016-10-24 05:13:06.734   
3 2016-10-24 05:14:17.192   
4 2016-10-24 05:14:24.625   

  Are you going actually going trick or treating yourself? Your gender:  \
0                                                 No               Male   
1                                                 No               Male   
2                                                 No             Female   
3                                                 No               Male   
4                                                Yes               Male   

  How old are you? Which country do you live in?  \
0               22                        Canada   
1               45                           usa   
2               48                            US   
3               57                           usa   
4               42                           USA   

  Which state, province, county do you live in?  [100 Grand Bar]  \
0                                       Ontario              JOY   
1                                            il              MEH   
2                                      Colorado              JOY   
3                                            il              JOY   
4                                  South Dakota              MEH   

   [Anonymous brown globs that come in black and orange wrappers]  \
0                                            DESPAIR                
1                                                MEH                
2                                            DESPAIR                
3                                                MEH                
4                                            DESPAIR                

   [Any full-sized candy bar]  [Black Jacks]  ...  \
0                         JOY            MEH  ...   
1                         JOY            JOY  ...   
2                         JOY            MEH  ...   
3                         JOY            MEH  ...   
4                         JOY        DESPAIR  ...   

  Please estimate the degree(s) of separation you have from the following celebrities [JK Rowling]  \
0                                        3 or higher                                                 
1                                        3 or higher                                                 
2                                        3 or higher                                                 
3                                        3 or higher                                                 
4                                        3 or higher                                                 

  Please estimate the degree(s) of separation you have from the following celebrities [JJ Abrams]  \
0                                                  2                                                
1                                        3 or higher                                                
2                                        3 or higher                                                
3                                        3 or higher                                                
4                                        3 or higher                                                

  Please estimate the degree(s) of separation you have from the following celebrities [Beyoncé]  \
0                                        3 or higher                                              
1                                        3 or higher                                              
2                                        3 or higher                                              
3                                        3 or higher                                              
4                                        3 or higher                                              

  Please estimate the degree(s) of separation you have from the following celebrities [Bieber]  \
0                    

Here I see that there aren't any issues with the headers other than the fact that they are very long. I see that the last column of the dataset has some problems because every piece of data in that column is coming in as NaN. There are some columns that I find very confusing, particularly the ones in brackets. Neither the answers or the column names themselves make sense. I see that there are many ways data can very in every column. Things can be spelled and written differently. For example, in the country column some people wrote "USA" while other wrote "US" and "usa". In the state cokums some people wrote the full state name and some only wrote the two letter symbol. There are many other places in the data where this could take place. Finally, I can tell that this data is about a lot of things, as the coulmns feature random questions about each person along with very reasonable question like age, gender, etc.

In [160]:
candy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1259 entries, 0 to 1258
Columns: 123 entries, Timestamp to  [York Peppermint Patties] Ignore
dtypes: datetime64[ns](1), float64(1), object(121)
memory usage: 1.2+ MB


This affirmed everything I stated earlier while also tellng me that this dataset has 129 rows and 123 columns.

In [161]:
candy.tail()

Timestamp  \
1254 2016-10-29 16:53:52.516   
1255 2016-10-30 06:53:54.735   
1256 2016-10-30 11:06:10.827   
1257 2016-10-30 16:07:26.539   
1258 2016-10-30 17:06:45.660   

     Are you going actually going trick or treating yourself? Your gender:  \
1254                                                 No             Female   
1255                                                 No               Male   
1256                                                 No               Male   
1257                                                 No               Male   
1258                                                Yes             Female   

     How old are you? Which country do you live in?  \
1254               52                           USA   
1255               33                 united states   
1256              NaN                           NaN   
1257               48                        canada   
1258               44                            Us   

     Which state, province, county do you live in?  [100 Grand Bar]  \
1254                                            TX              JOY   
1255                                     minnesota              JOY   
1256                                           NaN              JOY   
1257                                            BC              NaN   
1258                                            Nh              JOY   

      [Anonymous brown globs that come in black and orange wrappers]  \
1254                                            DESPAIR                
1255                                            DESPAIR                
1256                                                MEH                
1257                                            DESPAIR                
1258                                                MEH                

      [Any full-sized candy bar]  [Black Jacks]  ...  \
1254                         JOY            MEH  ...   
1255                         JOY        DESPAIR  ...   
1256                         JOY            NaN  ...   
1257                         JOY        DESPAIR  ...   
1258                         JOY            JOY  ...   

     Please estimate the degree(s) of separation you have from the following celebrities [JK Rowling]  \
1254                                        3 or higher                                                 
1255                               Actually, that's me.                                                 
1256                                                NaN                                                 
1257                                                  1                                                 
1258                                        3 or higher                                                 

     Please estimate the degree(s) of separation you have from the following celebrities [JJ Abrams]  \
1254                                        3 or higher                                                
1255                                        3 or higher                                                
1256                                                NaN                                                
1257                                                  2                                                
1258                                        3 or higher                                                

     Please estimate the degree(s) of separation you have from the following celebrities [Beyoncé]  \
1254                                        3 or higher                                              
1255                                        3 or higher                                              
1256                                                NaN                                              
1257                                        3 or higher                                              
1258                                        3 or higher        

Here I made sure there were no problems with the tail like the in the attendance data. There weren't any problems.

In [162]:
candy

Timestamp  \
0    2016-10-24 05:09:23.033   
1    2016-10-24 05:09:54.798   
2    2016-10-24 05:13:06.734   
3    2016-10-24 05:14:17.192   
4    2016-10-24 05:14:24.625   
5    2016-10-24 05:16:43.126   
6    2016-10-24 05:16:50.572   
7    2016-10-24 05:16:55.324   
8    2016-10-24 05:17:49.292   
9    2016-10-24 05:18:51.021   
10   2016-10-24 05:20:21.125   
11   2016-10-24 05:21:19.858   
12   2016-10-24 05:21:27.708   
13   2016-10-24 05:22:12.699   
14   2016-10-24 05:22:30.678   
15   2016-10-24 05:24:08.316   
16   2016-10-24 05:24:11.446   
17   2016-10-24 05:26:18.333   
18   2016-10-24 05:26:35.167   
19   2016-10-24 05:27:00.491   
20   2016-10-24 05:30:39.763   
21   2016-10-24 05:31:48.862   
22   2016-10-24 05:33:18.915   
23   2016-10-24 05:33:28.488   
24   2016-10-24 05:34:36.438   
25   2016-10-24 05:34:37.116   
26   2016-10-24 05:34:40.001   
27   2016-10-24 05:35:45.940   
28   2016-10-24 05:36:13.156   
29   2016-10-24 05:36:44.146   
...                      ...   
1229 2016-10-28 09:10:29.100   
1230 2016-10-28 09:10:38.669   
1231 2016-10-28 09:51:59.394   
1232 2016-10-28 10:12:14.405   
1233 2016-10-28 10:19:10.361   
1234 2016-10-28 10:32:54.679   
1235 2016-10-28 10:54:41.121   
1236 2016-10-28 12:41:25.987   
1237 2016-10-28 12:59:36.013   
1238 2016-10-28 13:08:58.202   
1239 2016-10-28 13:16:27.652   
1240 2016-10-28 13:42:50.055   
1241 2016-10-28 14:26:09.078   
1242 2016-10-28 14:37:24.766   
1243 2016-10-28 15:58:09.398   
1244 2016-10-28 17:05:16.492   
1245 2016-10-28 17:37:43.109   
1246 2016-10-28 19:45:24.355   
1247 2016-10-28 21:41:51.875   
1248 2016-10-29 00:55:44.688   
1249 2016-10-29 07:21:44.555   
1250 2016-10-29 10:53:26.543   
1251 2016-10-29 11:56:55.930   
1252 2016-10-29 12:02:52.773   
1253 2016-10-29 14:47:43.907   
1254 2016-10-29 16:53:52.516   
1255 2016-10-30 06:53:54.735   
1256 2016-10-30 11:06:10.827   
1257 2016-10-30 16:07:26.539   
1258 2016-10-30 17:06:45.660   

     Are you going actually going trick or treating yourself?  \
0                                                    No         
1                                                    No         
2                                                    No         
3                                                    No         
4                                                   Yes         
5                                                    No         
6                                                   Yes         
7                                                    No         
8                                                   Yes         
9                                                    No         
10                                                   No         
11                                                   No         
12                                                   No         
13                                                  Yes         
14                                                   No         
15                                                   No         
16                                                   No         
17                                                   No         
18                                                  Yes         
19                                                   No         
20                                                   No         
21                                                   No         
22                                                   No         
23                                                   No         
24                                                   No         
25                                                   No         
26                                                   No         
27                                                   No         
28                                                   No         
29                                                   No         
...                

Here I printed the whole dataset to quickly scroll through like the attendance dataset to make sure there were no rows with no data like the attendance dataset. I didn't find any rows without data, but I definatley encountered some of the problems I outlined above when I printed the head.

In [163]:
sales.head()

Unnamed: 0              event_id        primary_act_id  \
0           1  abcaf1adb99a935fc661  43f0436b905bfa7c2eec   
1           2  6c56d7f08c95f2aa453c  1a3e9aecd0617706a794   
2           3  c7ab4524a121f9d687d2  4b677c3f5bec71eec8d1   
3           4  394cb493f893be9b9ed1  b1ccea01ad6ef8522796   
4           5  55b5f67e618557929f48  91c03a34b562436efa3c   

       secondary_act_id   purch_party_lkup_id  \
0  b85143bf51323b72e53c  7dfa56dd7d5956b17587   
1  f53529c5679ea6ca5a48  4f9e6fc637eaf7b736c2   
2  b85143bf51323b72e53c  6c2545703bd527a7144d   
3  b85143bf51323b72e53c  527d6b1eaffc69ddd882   
4  b85143bf51323b72e53c  8bd62c394a35213bdf52   

                                          event_name  \
0  Xfinity Center Mansfield Premier Parking: Flor...   
1      Gorge Camping - dave matthews band - sept 3-7   
2        Dodge Theatre Adams Street Parking - benise   
3  Gexa Energy Pavilion Vip Parking : kid rock wi...   
4                      Premier Parking - motley crue   

                           primary_act_name  secondary_act_name  \
0  XFINITY Center Mansfield Premier Parking                 NaN   
1                             Gorge Camping  Dave Matthews Band   
2                             Parking Event                 NaN   
3          Gexa Energy Pavilion VIP Parking                 NaN   
4  White River Amphitheatre Premier Parking                 NaN   

  major_cat_name minor_cat_name  ... edu_1st_indv_val edu_2nd_indv_val  \
0           MISC        PARKING  ...              NaN              NaN   
1           MISC        CAMPING  ...              NaN              NaN   
2           MISC        PARKING  ...              NaN              NaN   
3           MISC        PARKING  ...              NaN              NaN   
4           MISC        PARKING  ...              NaN              NaN   

  adults_in_hh_num  married_ind  child_present_ind home_owner_ind occpn_val  \
0              NaN          NaN                NaN            NaN       NaN   
1              NaN          NaN                NaN            NaN       NaN   
2              NaN          NaN                NaN            NaN       NaN   
3              NaN          NaN                NaN            NaN       NaN   
4              NaN          NaN                NaN            NaN       NaN   

  occpn_1st_val occpn_2nd_val dist_to_ven  
0           NaN           NaN         NaN  
1           NaN           NaN        59.0  
2           NaN           NaN         NaN  
3           NaN           NaN         NaN  
4           NaN           NaN         NaN  

[5 rows x 46 columns]

This tells me that there is one unnamed column which features the row number. It isn't important for the rest of the data, so it can be removed later. I also see that though the rest of the columns have short names, none of them make sense given the fact that they don't provide much context. Finally, I see that there are a ton of NaNs throughout the dataset which will make it very hard to use.

In [164]:
sales.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 46 columns):
Unnamed: 0                5000 non-null int64
event_id                  5000 non-null object
primary_act_id            5000 non-null object
secondary_act_id          5000 non-null object
purch_party_lkup_id       5000 non-null object
event_name                5000 non-null object
primary_act_name          5000 non-null object
secondary_act_name        1586 non-null object
major_cat_name            5000 non-null object
minor_cat_name            5000 non-null object
la_event_type_cat         5000 non-null object
event_disp_name           5000 non-null object
ticket_text               5000 non-null object
tickets_purchased_qty     5000 non-null int64
trans_face_val_amt        5000 non-null float64
delivery_type_cd          5000 non-null object
event_date_time           5000 non-null object
event_dt                  5000 non-null object
presale_dt                2108 non-null object
o

This affirms what I said earlier and also tells me that there are 46 coumns and 5000 rows.

In [165]:
sales.tail()

Unnamed: 0              event_id        primary_act_id  \
4995        4996  79372bc1e7e14dfd3f82  a14232befff04be1e2f3   
4996        4997  f3b0de6a19e2a16e4fb1  83e5715f0ed3e13e6398   
4997        4998  7c9e2f6847a2dbe4ebff  9ea8ab40f58d18305ac8   
4998        4999  7af1ccd9a3ec3b6cef1c  e274b5c11a2a63e00141   
4999        5000  802f4bf8026e38c2acf2  24c326bd80bd2c8c9e17   

          secondary_act_id   purch_party_lkup_id  \
4995  b85143bf51323b72e53c  8a03312852bd2a785f2c   
4996  b2b558b1fe3efad19edb  f032fe520e8313bef880   
4997  b85143bf51323b72e53c  f270f4daa5c3dd61a8c5   
4998  b85143bf51323b72e53c  55a61a5e0ceb8beee897   
4999  b22d584a2132b33e23e0  55e6e08d4b712243944d   

                                             event_name  \
4995  Premier Parking - eric clapton - pnc bank arts...   
4996                     Sublime With Rome / rebelution   
4997  Premier Parking : journey with heart and cheap...   
4998                     Premier Parking: Rascal Flatts   
4999                                   Dropkick Murphys   

                                  primary_act_name secondary_act_name  \
4995  PNC Bank Arts Center Holmdel Premier Parking                NaN   
4996                             Sublime with Rome         Rebelution   
4997      Sleep Train Amphitheatre Premier Parking                NaN   
4998   Umb Bank Pavilion St. Louis Premier Parking                NaN   
4999                              Dropkick Murphys         Tiger Army   

     major_cat_name    minor_cat_name  ... edu_1st_indv_val edu_2nd_indv_val  \
4995           MISC           PARKING  ...              NaN              NaN   
4996       CONCERTS          ROCK/POP  ...              NaN              NaN   
4997           MISC           PARKING  ...              NaN              NaN   
4998           MISC           PARKING  ...              NaN              NaN   
4999       CONCERTS  ALTERNATIVE ROCK  ...              NaN              NaN   

     adults_in_hh_num  married_ind  child_present_ind home_owner_ind  \
4995              NaN          NaN                NaN            NaN   
4996              NaN          NaN                NaN            NaN   
4997              NaN          NaN                NaN            NaN   
4998              NaN          NaN                NaN            NaN   
4999              NaN          NaN                NaN            NaN   

     occpn_val occpn_1st_val occpn_2nd_val dist_to_ven  
4995       NaN           NaN           NaN         NaN  
4996       NaN           NaN           NaN         NaN  
4997       NaN           NaN           NaN         5.0  
4998       NaN           NaN           NaN         NaN  
4999       NaN           NaN           NaN         NaN  

[5 rows x 46 columns]

Here I made sure there were no problems with the tail like the in the attendance data. There weren't any problems.

In [166]:
sales

Unnamed: 0              event_id        primary_act_id  \
0              1  abcaf1adb99a935fc661  43f0436b905bfa7c2eec   
1              2  6c56d7f08c95f2aa453c  1a3e9aecd0617706a794   
2              3  c7ab4524a121f9d687d2  4b677c3f5bec71eec8d1   
3              4  394cb493f893be9b9ed1  b1ccea01ad6ef8522796   
4              5  55b5f67e618557929f48  91c03a34b562436efa3c   
5              6  4f10fd8b9f550352bd56  ac4b847b3fde66f2117e   
6              7  105ea9ee10065b54b0df  a14232befff04be1e2f3   
7              8  593b910d17913d7101cb  91c03a34b562436efa3c   
8              9  0fb4e7152e09a40ffedd  0efaba7ce3f0d7466b42   
9             10  93fce8094d89e0a94aac  f6425a3223e73ea6de5a   
10            11  402bebb1d89c4024047a  8f53c0b7cd7810a1829d   
11            12  6adcf8b2dc8a761b22b7  094156ef72361776c8ee   
12            13  d3c8f5fb844690fb5af6  732d4f52a3ac2b7de8ef   
13            14  5c7d36b783905ba488f4  34ad0e9254b8f7d1cbe0   
14            15  e68754d15ef2b8694b26  6e3df9919c998817e1b1   
15            16  afd7dbe7675fc5480a0c  4b677c3f5bec71eec8d1   
16            17  e6a1bc0620dc1fb45b6d  7cb02562bcb9387cbd13   
17            18  8bc575c22b944a6715d0  43f0436b905bfa7c2eec   
18            19  5777fc7899ca8f4ef03b  4b677c3f5bec71eec8d1   
19            20  b1574cb6bfd591a98572  d769ab09a96be5a0b316   
20            21  516c79412f8b76efa1ac  43f0436b905bfa7c2eec   
21            22  f7a4ec7efe1cd5125812  ac4b847b3fde66f2117e   
22            23  de38ce6ac8d362f98fd7  6cdc2e270775b7e2f709   
23            24  179eb086e65ad82d0225  9caaa3cea047cfffa7b3   
24            25  7507d55fcb3b1c524027  dce08e99e3ce7d2149f3   
25            26  6c73e7adb3a2982b1456  6cdc2e270775b7e2f709   
26            27  dcc303578fdb9493de39  a75e1e1b72ac7f7ab1f6   
27            28  5ee2a492f932af193757  be462316a8567e40fdf0   
28            29  dfe06d7bf50a876cc303  4b677c3f5bec71eec8d1   
29            30  b516196d16d9f600ed69  8475a7f6159d0149c11d   
...          ...                   ...                   ...   
4970        4971  a4f431802a5db1a5afee  86548c5bce2089a5d311   
4971        4972  ffdb7f76c259bd015a54  ac4b847b3fde66f2117e   
4972        4973  7eca92b4b4eb41ad6239  1a3e9aecd0617706a794   
4973        4974  d09d372bed59ac89a735  002d7347ccac23f064b3   
4974        4975  6dc00a98a31de0cf10ac  c2771bc344c9db424eac   
4975        4976  2121b4186f9d0232aae1  c630e0584b7b59c4edbe   
4976        4977  1bc30cba19894b1a3513  9babfdb02a7f8004fc6e   
4977        4978  7071ee5754b7ed08b0ef  ea6991e953477c0bb31f   
4978        4979  b7926005e963004ec487  e7977ac6f61f6a9500ab   
4979        4980  21c735bbf5a29ee1db09  55503bb8973acb4c0f0a   
4980        4981  f1b4add9237241bd69c4  8878d2a6d6d6a71ca7a2   
4981        4982  6a29cec2d939911d4c1c  d0b2f8ad63a0301c1d78   
4982        4983  4479d1e5eab4b6703142  c1a7fb9a4de3521c4b87   
4983        4984  3af4addbf8aa67f62e6b  711a6302877f7e3028b9   
4984        4985  84a260b1bcd31e2e75a7  1a3e9aecd0617706a794   
4985        4986  9bb699d6f5b40789356e  bb5dacc7476668d0dd50   
4986        4987  445c9e36e356c2995a73  5cd02f0925d7ba013e34   
4987        4988  474219cfd57afb7210ef  6e3df9919c998817e1b1   
4988        4989  daa12dc215f196c24bc1  8d43f8055cd8801eee81   
4989        4990  893e8bc79ada0d860ac9  ebe30b90662305cd8f18   
4990        4991  6789e447a44036d5628c  4b677c3f5bec71eec8d1   
4991        4992  09984e0d6ca9b1de13ed  da9ab9b6f5d99b8c3195   
4992        4993  720f557e40ffdb2cfa16  91c03a34b562436efa3c   
4993        4994  dfbc00d338c1e2e99d4b  a0ed82451f57a8914eba   
4994        4995  997c44d1def48153316d  716df6554f0887b73f1a   
4995        4996  79372bc1e7e14dfd3f82  a14232befff04be1e2f3   
4996        4997  f3b0de6a19e2a16e4fb1  83e5715f0ed3e13e6398   
4997        4998  7c9e2f6847a2dbe4ebff  9ea8ab40f58d18305ac8   
4998        4999  7af1ccd9a3ec3b6cef1c  e274b5c11a2a63e00141   
4999        5000  802f4bf8026e38c2acf2  24c326bd80bd2c8c9e17   

          secondary_act_id   purch_pa

Here I printed the whole dataset to quickly scroll through like the attendance dataset to make sure there were no rows with no data like the attendance dataset. I didn't find any rows without data, but I found that the rows I had previously believed to have had no data actually all had some data, meaning this dataset doesn't have a ton of cleaning needed.

I have decided to clean the attendence and sales datasets. I chose to clean these two datasets because they both seem to have very different issues that seem like they will be more complicated and therefor more useful to practice for future projects. By doing both of them I develope two very different skills while candy is pretty similar to both. The attendance dataset has horrible issues stemming from the headers and also has some rows without data while the sales dataset has a bad column along with a lot of NaN. Though the sales dataset has these probelms it doesn't have nearly as severe problems, so I choose it because it will help me to learn how to clean every aspect of a dataset that I need to even when it is virtually clean already.

### Dataset Preparation 1 (Attendence)

In [167]:
labels = ["Country", "E + S ADA as % enrollment 1", "E + S ADA as % enrollment 2", "E + S average hours in day 1", "E + S average hours in day 2", "E + S average days in year 1", "E + S average days in year 2", "E + S average hours in year 1", "E + S average hours in year 2", "E ADA as % enrollment 1", "E ADA as % enrollment 2", "E average hours in day 1", "E average hours in day 2","S ADA as % enrollment 1", "S ADA as % enrollment 2", "S average hours in day 1", "S average hours in day 2" ]
attendance1 = pd.read_excel('attendance.xls', encoding = "iso-8859-1",
                  header=3,
                  names=labels)
index_col=False
attendance1.head()

Country  E + S ADA as % enrollment 1  \
0     United States ........                    93.078962   
1  Alabama .................                    93.812371   
2  Alaska ..................                    89.917597   
3  Arizona .................                    89.036961   
4  Arkansas ................                    91.827111   

   E + S ADA as % enrollment 2  E + S average hours in day 1  \
0                     0.218555                       6.64470   
1                     1.237261                       7.02852   
2                     1.215976                       6.47688   
3                     2.952601                       6.43369   
4                     1.352924                       6.88542   

   E + S average hours in day 2  E + S average days in year 1  \
0                      0.017650                         180.0   
1                      0.065621                         180.0   
2                      0.049882                         180.0   
3                      0.091917                         181.0   
4                      0.061468                         179.0   

   E + S average days in year 2  E + S average hours in year 1  \
0                      0.142993                      1192.6472   
1                      0.755110                      1266.6205   
2                      3.427735                      1162.9084   
3                      1.678664                      1159.1144   
4                      0.201577                      1228.8881   

   E + S average hours in year 2  E ADA as % enrollment 1  \
0                         3.0859                94.004983   
1                        12.3343                93.776375   
2                        22.8639                91.281537   
3                        14.4327                88.904393   
4                        10.6821                92.093201   

   E ADA as % enrollment 2  E average hours in day 1  \
0                 0.269369                  6.656056   
1                 1.839154                  7.038473   
2                 1.563482                  6.486785   
3                 3.905033                  6.443155   
4                 2.091651                  6.932160   

   E average hours in day 2 S ADA as % enrollment 1 S ADA as % enrollment 2  \
0                  0.015998                 91.1181                0.432223   
1                  0.075944                  94.561                 0.37928   
2                  0.053082                 93.2409                 1.57371   
3                  0.102053                 88.9797                 3.22058   
4                  0.076546                  90.812                 2.23155   

  S average hours in day 1 S average hours in day 2  
0                  6.59439                 0.040308  
1                  7.13724                  0.17312  
2                  6.24087                 0.145993  
3                  6.36909                 0.248824  
4                  6.76751                 0.095928

Here I replaced the headers to make it only one row of headers. This made it so the dataset is much easier to understand and got rid of almost all of the NaNs that were present because of the weird headers. I made this a new variable so it wouldn't impact the results I get in the dataset exploration section. On the first line I wrote the new headers. Every original header contained two columns of data, so I ended up making two headers for each piece of data to get every column filled. On the second line I had pandad read the dataset, on the third line I states that rows 0, 1, 2, and 3 were header rows that needed to be deleted, and on the fourth line I replaced the headers with the ones I wrote. I then added a piece of code that I will need in order to remove certain rows later on on line five. Finally, I printed the head of the dataset to make sure it worked, and it did. The row three (the fourth row) wasn't a header, but something that labeled each section of the dataset with a number. This was also creating some problems so I added it to the list of headers to get rid of it.

In [168]:
attendance1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65 entries, 0 to 64
Data columns (total 17 columns):
Country                          56 non-null object
E + S ADA as % enrollment 1      52 non-null float64
E + S ADA as % enrollment 2      52 non-null float64
E + S average hours in day 1     52 non-null float64
E + S average hours in day 2     52 non-null float64
E + S average days in year 1     52 non-null float64
E + S average days in year 2     52 non-null float64
E + S average hours in year 1    52 non-null float64
E + S average hours in year 2    52 non-null float64
E ADA as % enrollment 1          52 non-null float64
E ADA as % enrollment 2          52 non-null float64
E average hours in day 1         52 non-null float64
E average hours in day 2         52 non-null float64
S ADA as % enrollment 1          52 non-null object
S ADA as % enrollment 2          52 non-null object
S average hours in day 1         52 non-null object
S average hours in day 2         52 non-null object
d

Here I double-checked to make sure I got everything right by using info intead of head, and everything was right.

In [169]:
attendance1.drop([61], inplace=True)
attendance1.drop([62], inplace=True)
attendance1.drop([63], inplace=True)
attendance1.drop([64], inplace=True)
attendance1.tail()

Country  E + S ADA as % enrollment 1  \
56  Virginia ................                    94.677993   
57  Washington ..............                    82.855625   
58  West Virginia ...........                    94.009760   
59  Wisconsin ...............                    94.950856   
60  Wyoming .................                    92.353008   

    E + S ADA as % enrollment 2  E + S average hours in day 1  \
56                     0.460467                       6.61718   
57                     3.056990                       6.22202   
58                     0.987722                       6.87364   
59                     0.566144                       6.91175   
60                     1.154487                       6.85867   

    E + S average hours in day 2  E + S average days in year 1  \
56                      0.050234                         181.0   
57                      0.080353                         179.0   
58                      0.070124                         182.0   
59                      0.042731                         180.0   
60                      0.045848                         175.0   

    E + S average days in year 2  E + S average hours in year 1  \
56                      0.159185                      1197.3283   
57                      0.173674                      1116.5367   
58                      0.497818                      1251.2268   
59                      0.736128                      1246.4809   
60                      1.278829                      1200.9323   

    E + S average hours in year 2  E ADA as % enrollment 1  \
56                         9.1034                95.596850   
57                        14.4906                81.047004   
58                        13.6853                94.374244   
59                         8.6319                95.375952   
60                         8.3266                92.190909   

    E ADA as % enrollment 2  E average hours in day 1  \
56                 0.349639                  6.611134   
57                 4.416025                  6.378638   
58                 1.315128                  6.879590   
59                 0.414753                  6.886868   
60                 1.647681                  6.850612   

    E average hours in day 2 S ADA as % enrollment 1 S ADA as % enrollment 2  \
56                  0.032699                 93.9422                  0.5841   
57                  0.045202                 85.7672                 2.75371   
58                  0.093158                  92.846                0.408932   
59                  0.048538                 93.0196                 1.90793   
60                  0.054651                 92.3894                0.749597   

   S average hours in day 1 S average hours in day 2  
56                  6.67654                 0.187479  
57                  5.89289                 0.266992  
58                  6.82534                 0.136584  
59                  7.01951                 0.144539  
60                  6.97859                   0.0679

Here I removed the last four rows of the dataset to get rid of the problems involving the tail. This way I can get rid of the NaNs that were present. I then printed the dataset to make sure it worked. It worked well.

In [170]:
attendance1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 61 entries, 0 to 60
Data columns (total 17 columns):
Country                          52 non-null object
E + S ADA as % enrollment 1      52 non-null float64
E + S ADA as % enrollment 2      52 non-null float64
E + S average hours in day 1     52 non-null float64
E + S average hours in day 2     52 non-null float64
E + S average days in year 1     52 non-null float64
E + S average days in year 2     52 non-null float64
E + S average hours in year 1    52 non-null float64
E + S average hours in year 2    52 non-null float64
E ADA as % enrollment 1          52 non-null float64
E ADA as % enrollment 2          52 non-null float64
E average hours in day 1         52 non-null float64
E average hours in day 2         52 non-null float64
S ADA as % enrollment 1          52 non-null object
S ADA as % enrollment 2          52 non-null object
S average hours in day 1         52 non-null object
S average hours in day 2         52 non-null object
d

Here I used .info to double check and make sure everything worked right and it did.

In [171]:
attendance1.drop([6], inplace=True)
attendance1.drop([12], inplace=True)
attendance1.drop([18], inplace=True)
attendance1.drop([24], inplace=True)
attendance1.drop([30], inplace=True)
attendance1.drop([36], inplace=True)
attendance1.drop([42], inplace=True)
attendance1.drop([48], inplace=True)
attendance1.drop([54], inplace=True)
attendance1

Country  E + S ADA as % enrollment 1  \
0      United States ........                    93.078962   
1   Alabama .................                    93.812371   
2   Alaska ..................                    89.917597   
3   Arizona .................                    89.036961   
4   Arkansas ................                    91.827111   
5   California ..............                    93.241017   
7   Colorado ................                    93.887162   
8   Connecticut .............                    87.925871   
9   Delaware ................                    89.770581   
10  District of Columbia ....                    91.220829   
11  Florida .................                    92.678174   
13  Georgia .................                    93.266158   
14  Hawaii ..................                    90.721279   
15  Idaho ...................                    92.373340   
16  Illinois ................                    93.970461   
17  Indiana .................                    95.674644   
19  Iowa ....................                    94.760592   
20  Kansas ..................                    95.350826   
21  Kentucky ................                    93.068117   
22  Louisiana ...............                    90.349523   
23  Maine ...................                    90.303632   
25  Maryland ................                    94.103682   
26  Massachusetts ...........                    94.596304   
27  Michigan ................                    92.968459   
28  Minnesota ...............                    93.146022   
29  Mississippi .............                    92.091106   
31  Missouri ................                    94.823437   
32  Montana .................                    91.325266   
33  Nebraska ................                    94.944186   
34  Nevada ..................                    93.515077   
35  New Hampshire ...........                    92.207444   
37  New Jersey ..............                    94.610399   
38  New Mexico ..............                    91.873346   
39  New York ................                    92.735377   
40  North Carolina ..........                    92.593062   
41  North Dakota ............                    95.921212   
43  Ohio ....................                    91.780603   
44  Oklahoma ................                    92.057487   
45  Oregon ..................                    94.401409   
46  Pennsylvania ............                    94.880593   
47  Rhode Island ............                    93.736534   
49  South Carolina ..........                    94.897206   
50  South Dakota ............                    93.640576   
51  Tennessee ...............                    94.942545   
52  Texas ...................                    94.086751   
53  Utah ....................                    91.367433   
55  Vermont .................                    92.726006   
56  Virginia ................                    94.677993   
57  Washington ..............                    82.855625   
58  West Virginia ...........                    94.009760   
59  Wisconsin ...............                    94.950856   
60  Wyoming .................                    92.353008   

    E + S ADA as % enrollment 2  E + S average hours in day 1  \
0                      0.218555                       6.64470   
1                      1.237261                       7.02852   
2                      1.215976                       6.47688   
3                      2.952601                       6.43369   
4                      1.352924                       6.88542   
5                      0.711649                       6.24064   
7                      0.440366                       7.00761   
8                      2.984395                       6.46655   
9                      1.754942                       6.67801   
10                     1.269699                       6.91375   
11                  

Here I removed all of the rows that had no data. This got rid of all the rest of the NaNs. I then printed to make sure everything was right.

In [172]:
attendance1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 52 entries, 0 to 60
Data columns (total 17 columns):
Country                          52 non-null object
E + S ADA as % enrollment 1      52 non-null float64
E + S ADA as % enrollment 2      52 non-null float64
E + S average hours in day 1     52 non-null float64
E + S average hours in day 2     52 non-null float64
E + S average days in year 1     52 non-null float64
E + S average days in year 2     52 non-null float64
E + S average hours in year 1    52 non-null float64
E + S average hours in year 2    52 non-null float64
E ADA as % enrollment 1          52 non-null float64
E ADA as % enrollment 2          52 non-null float64
E average hours in day 1         52 non-null float64
E average hours in day 2         52 non-null float64
S ADA as % enrollment 1          52 non-null object
S ADA as % enrollment 2          52 non-null object
S average hours in day 1         52 non-null object
S average hours in day 2         52 non-null object
d

Here I used .info to double check and make sure everything worked right and it did.

### Dataset Preparation 2 (Sales)

In [173]:
sales1 = pd.read_csv('sales.csv', encoding = "iso-8859-1")

Here I created a new varible for the sales dataset so my cleaning won't affect the dataset exploration of this lab where I looked at the uncleaned datasets.

In [174]:
del sales1['Unnamed: 0']
sales1.head()

event_id        primary_act_id      secondary_act_id  \
0  abcaf1adb99a935fc661  43f0436b905bfa7c2eec  b85143bf51323b72e53c   
1  6c56d7f08c95f2aa453c  1a3e9aecd0617706a794  f53529c5679ea6ca5a48   
2  c7ab4524a121f9d687d2  4b677c3f5bec71eec8d1  b85143bf51323b72e53c   
3  394cb493f893be9b9ed1  b1ccea01ad6ef8522796  b85143bf51323b72e53c   
4  55b5f67e618557929f48  91c03a34b562436efa3c  b85143bf51323b72e53c   

    purch_party_lkup_id                                         event_name  \
0  7dfa56dd7d5956b17587  Xfinity Center Mansfield Premier Parking: Flor...   
1  4f9e6fc637eaf7b736c2      Gorge Camping - dave matthews band - sept 3-7   
2  6c2545703bd527a7144d        Dodge Theatre Adams Street Parking - benise   
3  527d6b1eaffc69ddd882  Gexa Energy Pavilion Vip Parking : kid rock wi...   
4  8bd62c394a35213bdf52                      Premier Parking - motley crue   

                           primary_act_name  secondary_act_name  \
0  XFINITY Center Mansfield Premier Parking                 NaN   
1                             Gorge Camping  Dave Matthews Band   
2                             Parking Event                 NaN   
3          Gexa Energy Pavilion VIP Parking                 NaN   
4  White River Amphitheatre Premier Parking                 NaN   

  major_cat_name minor_cat_name la_event_type_cat  ... edu_1st_indv_val  \
0           MISC        PARKING           PARKING  ...              NaN   
1           MISC        CAMPING           INVALID  ...              NaN   
2           MISC        PARKING           PARKING  ...              NaN   
3           MISC        PARKING           PARKING  ...              NaN   
4           MISC        PARKING           PARKING  ...              NaN   

  edu_2nd_indv_val  adults_in_hh_num  married_ind child_present_ind  \
0              NaN               NaN          NaN               NaN   
1              NaN               NaN          NaN               NaN   
2              NaN               NaN          NaN               NaN   
3              NaN               NaN          NaN               NaN   
4              NaN               NaN          NaN               NaN   

  home_owner_ind occpn_val occpn_1st_val occpn_2nd_val dist_to_ven  
0            NaN       NaN           NaN           NaN         NaN  
1            NaN       NaN           NaN           NaN        59.0  
2            NaN       NaN           NaN           NaN         NaN  
3            NaN       NaN           NaN           NaN         NaN  
4            NaN       NaN           NaN           NaN         NaN  

[5 rows x 45 columns]

Here I cleaned my first column of the dataset by completely removing the first column. This column just repeated the row numbers but was one off and didn't have a header, so it was best just to get rid of it to make the dataset cleaner. I then printed the head of the dataset to make sure it worked and it did.

In [175]:
sales1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 45 columns):
event_id                  5000 non-null object
primary_act_id            5000 non-null object
secondary_act_id          5000 non-null object
purch_party_lkup_id       5000 non-null object
event_name                5000 non-null object
primary_act_name          5000 non-null object
secondary_act_name        1586 non-null object
major_cat_name            5000 non-null object
minor_cat_name            5000 non-null object
la_event_type_cat         5000 non-null object
event_disp_name           5000 non-null object
ticket_text               5000 non-null object
tickets_purchased_qty     5000 non-null int64
trans_face_val_amt        5000 non-null float64
delivery_type_cd          5000 non-null object
event_date_time           5000 non-null object
event_dt                  5000 non-null object
presale_dt                2108 non-null object
onsale_dt                 4899 non-null object


Here I used .info to double check and make sure everything worked right and it did.

In [176]:
sales1.describe()

tickets_purchased_qty  trans_face_val_amt      age_yr     income_amt  \
count             5000.00000         5000.000000  168.000000     154.000000   
mean                 1.63900           77.077446   44.369048   86396.103896   
std                  1.11361          123.864788   16.531067   58895.045547   
min                  1.00000            1.000000   18.000000   10000.000000   
25%                  1.00000           20.000000   30.000000   45000.000000   
50%                  1.00000           30.000000   44.000000   62500.000000   
75%                  2.00000           85.000000   56.000000   87500.000000   
max                  8.00000         1520.880000   94.000000  200000.000000   

       adults_in_hh_num  married_ind  child_present_ind  home_owner_ind  \
count        176.000000   157.000000         130.000000      138.000000   
mean           2.579545     0.668790           0.561538        0.942029   
std            1.490707     0.472155           0.498118        0.234540   
min            1.000000     0.000000           0.000000        0.000000   
25%            1.000000     0.000000           0.000000        1.000000   
50%            2.000000     1.000000           1.000000        1.000000   
75%            4.000000     1.000000           1.000000        1.000000   
max            6.000000     1.000000           1.000000        1.000000   

       dist_to_ven  
count   323.000000  
mean    158.188854  
std     395.678376  
min       0.000000  
25%      12.000000  
50%      26.000000  
75%      77.500000  
max    2548.000000

In [177]:
del sales1['event_date_time']
sales1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 44 columns):
event_id                  5000 non-null object
primary_act_id            5000 non-null object
secondary_act_id          5000 non-null object
purch_party_lkup_id       5000 non-null object
event_name                5000 non-null object
primary_act_name          5000 non-null object
secondary_act_name        1586 non-null object
major_cat_name            5000 non-null object
minor_cat_name            5000 non-null object
la_event_type_cat         5000 non-null object
event_disp_name           5000 non-null object
ticket_text               5000 non-null object
tickets_purchased_qty     5000 non-null int64
trans_face_val_amt        5000 non-null float64
delivery_type_cd          5000 non-null object
event_dt                  5000 non-null object
presale_dt                2108 non-null object
onsale_dt                 4899 non-null object
sales_ord_create_dttm     4996 non-null object


Here I cleaned my second coulumn of data. There was a column in the middle of the dataset that contained no data whatsoever, so I deleted it because it was useless. This column was burried in the dataset, so I couldn't see it when printing the head so I just looked at the info to see if it worked, and it did because the number of columns went from 45 to 44.

### Conclusion

I felt as if the attendance dataset was much less clean than the sales dataset. I found that the attenance dataset didn't have many problems in the data itself, but more problems in the way it was formatted. First of all, there were three lines of headers. Second, each header contained two columns of data, and third there were many rows that had no data at all and were merely there as dividers. Finally, at the end of the dataset there was a thing of notes that needed to be gotten rid of. I feel as if I did a great job cleaning this dataset because I was able to fix all of these problems and make the dataset virtually completely clean. THe sales dataset had a lot fewer issues. The first column had a repeat of therow numbers, but it was off by one. Also, there were a lot of NaNs but those didn't make the dataset unclean they just demonstrated the fact that there was no data for the particular section. Finally, there was one column in which there was no data at all. Other than that, however, there were no issues. This was a very useable dataset even if I wasn't to clean it at all. I feel as if I did a very good job cleaning this dataset because even though it didn't have many issues I was able to fix all of the problems and make it incredibly clean.

### Acknowledgements

I would like to thank Ms. Sconyers and this website: https://stackoverflow.com/questions/13411544/delete-column-from-pandas-dataframe. They both helped me find code that I used to clean the datasets.